src/managed/Microsoft.DotNet.Build.Bundle/BundleManifest.cs

```diff
@@ -0,0 +1,208 @@
+        }
+
+        public FileEntry AddEntry(FileType type, string name, long offset, long size)
+        {
+            switch(type)
+            {
+                case FileType.DepsJson:
+                    flags |= BundleFlags.HasDepsJson;
+                    break;
+                case FileType.RuntimeConfigJson:
+                    flags |= BundleFlags.HasRuntimeConfigJson;
+                    break;
+                case FileType.RuntimeConfigDevJson:
+                    flags |= BundleFlags.HasRuntimeConfigDevJson;
+                    break;
+
+                default:
+                    break;
+            }
+
+            FileEntry entry = new FileEntry(type, name, offset, size);
+            Files.Add(entry);
+            return entry;
+        }
+
+        void WriteFileEntries(BinaryWriter writer)
+        {
+            Action<FileType> writeEntry = (FileType type) =>
+                Files
+                .Where(entry => entry.Type == type)
+                .Single()
+                .Write(writer);
+
+            Func<FileType, int> writeEntries = (FileType type) =>
+            {
+                IEnumerable<FileEntry> entries = Files.Where(entry => entry.Type == type);
+                foreach (FileEntry entry in entries)
+                    entry.Write(writer);
+                return entries.Count();
+            };
+
+            FileEntryStart = writer.BaseStream.Position;
+
+            if (flags.HasFlag(BundleFlags.HasDepsJson))
+                writeEntry(FileType.DepsJson);
+
+            if (flags.HasFlag(BundleFlags.HasRuntimeConfigJson))
+                writeEntry(FileType.RuntimeConfigJson);
+
+            if (flags.HasFlag(BundleFlags.HasRuntimeConfigDevJson))
+                writeEntry(FileType.RuntimeConfigDevJson);
+
+            writeEntry(FileType.Application);
+
+            // Count is +1 For the main app already written.
+            AssemblyCount = writeEntries(FileType.Assembly) + 1;
+            writeEntries(FileType.Other);
+        }
+
+        public void Write(BinaryWriter writer)
+        {
+            long startOffset = writer.BaseStream.Position;
+            WriteFileEntries(writer);
+
+            writer.Write(MajorVersion);
+            writer.Write(MinorVersion);
+
+            writer.Write(FileEntryStart);
+            writer.Write(Files.Count());
+            writer.Write(AssemblyCount);
+            writer.Write((int)flags);
+
+            writer.Write(Signature);
+
+            long size = writer.BaseStream.Position - startOffset;
+            Program.Log($"Manifest: Offset={startOffset}, Size={size}");
+        }
+
+        void ReadFileEntries(BinaryReader reader, long fileCount)
+        {
+            // This extractor doesn't actually care about the type of file.
+            // It just extracts out all files. 
+            reader.BaseStream.Position = FileEntryStart;
+            for (long i = 0; i < fileCount; i++)
+                Files.Add(FileEntry.Read(reader));
```

Comment:
Not sure if we have style guide for C# in this repo. Personally I prefer always using curly braces even for one-liners...